model/common/src/icon4py/model/common/grid/vertical.py

```diff
@@ -517,3 +539,141 @@ def get_vct_a_and_vct_b(vertical_config: VerticalGridConfig) -> tuple[fa.KField,
         if vertical_config.file_path
         else _compute_vct_a_and_vct_b(vertical_config)
     )
+
+
+def init_vert_coord(
+    vct_a: fa.KField[ta.wpfloat],
+    topography: fa.CellField[ta.wpfloat],
+    topography_smoothed: fa.CellField[ta.wpfloat],
+    grid: icon_grid.IconGrid,
+    vertical_config: VerticalGridConfig,
+    vertical_geometry: VerticalGrid,
+    nshift: int = 0,  # shift parameter used for vertical nesting
+) -> fa.CellKField[ta.wpfloat]:
+    """
+    Computes the 3D vertical coordinate fields.
+    """
+
+    def decay_func(vct_a, model_top_height, decay_scale, decay_exponent):
+        return xp.sinh(
+            (model_top_height / decay_scale) ** decay_exponent
+            - (vct_a / decay_scale) ** decay_exponent
+        ) / xp.sinh((model_top_height / decay_scale) ** decay_exponent)
+
+    z3d_i = xp.zeros((grid.num_cells, grid.num_levels + 1), dtype=ta.wpfloat)
+
+    topography = topography.asnumpy()
```

Comment:
Whenever you do a x.asnumpy() on a field, it will return the underlying CPU buffer as a numpy array if on CPU, copy from device to host if on GPU. So if you do this, then you might as well not use `xp.... below but do the computation explicitly in numpy. If you want the computation to be done on the same DEVICE that the field has been allocated then you can do a `x.ndarray` that is either the CPU or the GPU buffer and could then be processed with cupy.

Reply:
done, I did it in some other places but forgot here

Reply:
thanks for the explanation :-)